<a href="https://colab.research.google.com/github/sally20921/running-tests-in-computer-vision/blob/main/Google_Colab_Short_Tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Useful Resources
- [Google Colab Tutorial](https://colab.research.google.com/drive/16pBJQePbqkz3QFV54L4NIkOn1kwpuRrj)

# Virtual Machine

The most powerful features of google colab is the ability to use cloud GPU for free. Like the other destkop environment you can also access most of the bash command with a `!` added in the front of the command.

if you want to run a single script, you can use `!sh example.sh`.

At first, turn on the GPU from `Runtime -> Change Runtime Type -> Hardware Acceleration`.

The entire colab runs in a cloud VM (Virtual Machine). You will see that the current colab notebook is running on top of `Ubuntu 18.04.3 LTS` (at the time of this writing).

In [ ]:
!ls -l
#list files and directories in long format
# each file or directory, permissions, owner, group, size, modified date

total 4
drwxr-xr-x 1 root root 4096 Jan 11 17:02 sample_data


In [ ]:
!pwd
# print working directory: the full pathname of the current working directory

/content


In [ ]:
!ls -l /
# the forward slash(/) represents the root directory.
# lists the contents of the root directory in a long format

In [ ]:
!cat /etc/*release
# the `cat` command writes file contents to the standard output.
# `/etc/*release` matches any file in the `/etc` directory that ends with `release.`
# `cat /etc/*release` often contain information about the distribution and its version.

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy


# GPU Details

The GPU details can be accessed by `!nvidia-smi`.

In [ ]:
!nvidia-smi

Tue Jan 16 03:03:53 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   61C    P8              10W /  70W |      0MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

## Current Resources Info
- Python 3 Google Compute Engine backend (GPU)
- Showing resources from 12:03PM to 12:05PM
 - System RAM: $0.9/12.7$ GB
 - GPU RAM: None
 - Disk: $26.3/78.2$ GB

- NVIDIA SMI
 - Driver Version: 535.104.05
 - CUDA Version: 12.2
 - GPU Name: Tesla T4
 - Memory Usage: 0MiB/15360MiB


# Ubuntu 22.04.3 LTS and Drivers

Now that, as you are using Ubuntu, you can use any command that you use from GNU terminal. This makes life very simple.

**However, for your project, you have to find a way to transfer or download your data files here in this Virtual Machine.**
- The easiest way is to mount your google drive here in this VM and use git for version control of your codes.
- Another easiest way is to get yoru file is to use `wget` from a file server or dropbox.

To install a python package with `pip`, use the following. You can also use `apt-get` to install any package in ubuntu.

In [ ]:
!pip install gpustat
!pip install fairseq
!apt-get install build-essential

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.1/98.1 kB 2.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.4/58.4 kB 5.3 MB/s eta 0:00:00
  Created wheel for gpustat: filename=gpustat-1.1.1-py3-none-any.whl size=26535 sha256=b43996a1202743d77c849f0947d87278080311e045fc6825be7373b5f9970536
  Stored in directory: /root/.cache/pip/wheels/ec/d7/80/a71ba3540900e1f276bcae685efd8e590c810d2108b95f1e47
Successfully built gpustat
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 33.0 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 123.8/123.8 kB 19.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.3/106.3 kB 17.7 MB/s eta 0:00

In [ ]:
!pip install gpustat
! gpustat

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.1/98.1 kB 2.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.4/58.4 kB 8.3 MB/s eta 0:00:00
  Created wheel for gpustat: filename=gpustat-1.1.1-py3-none-any.whl size=26535 sha256=a909680b105f0af9a3ad889ed985a613744c2b31c7e98da104996bb9b8591e09
  Stored in directory: /root/.cache/pip/wheels/ec/d7/80/a71ba3540900e1f276bcae685efd8e590c810d2108b95f1e47
Successfully built gpustat
c9639a16b986  Thu Jan 18 01:52:36 2024  535.104.05
[0] Tesla T4 | 57°C,   0 % |     0 / 15360 MB |


Please check the nvidia driver version, and cuda version in current OS.

Not all versions support the latest cuda, cudnn, etc.
- Current Driver Version: 535.104.05
- Current CUDA Version: 12.2
- Python Version: 3.10.12
- PyTorch Version: 2.1.0+cu121
- PyTorch cuDNN Version: 8902

Make sure you can install your preferred deep learning library with the current nvidia-driver, cuda and cudnn.



---

📂 More about CUDA, cuDNN
- CUDA: a parallel computing platform and programming model for general computing on GPU
- cuDNN: GPU-accelerated library of primitives for deep neural network, **built on top of the CUDA framework**

| CUDA                                                                | cuDNN                                                                                                                                            |
|---------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------|
| general-purpose platform for parallel computing on GPUs             | specifically tailored to accelerate deep learning tasks on GPUs                                                                                  |
| speed up computing applications by harnessing the power of GPUs     | highly tuned implementations for standard routines in deep learning,  such as forward and backward attention, matmul, pooling and normalization. |
| lower-level platform for general-purpose parallel computing on GPUs | higher-level library optimized for deep learning tasks, built on top of CUDA framework                                                           |



---

☝ More about CPU, GPU, TPU and the testing environment

- CPU: often compared to the "brains" of a device, is responsible for executing instructions for programs from your system's memory.



In [ ]:
from platform import python_version
import torch

print("Python Version", python_version())
print("Pytorch Version", torch.__version__)
print("Pytorch cuDNN Version", torch.backends.cudnn.version())

Python Version 3.10.12
Pytorch Version 2.1.0+cu121
Pytorch cuDNN Version 8902


In [ ]:
!nvidia-smi

Thu Jan 18 01:53:37 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   54C    P8              10W /  70W |      0MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

# Uploading Data to Virtual Machine

There are largely three ways to upload existing data to Google Colab Virtual Machine.

1. Upload From Machine: explore the built-in left panel of Google Colab
2. Mount Google Drive: explore the built-in left panel of Google Colab
3. using `wget`
4. using `git clone` to clone git repository

### Mount Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!pwd
!ls

/content
drive  sample_data


### Using `wget`

In [4]:
!pwd
!wget https://dl.fbaipublicfiles.com/XML/codes_xnli_100
!ls -l
!head -n 10 codes_xnli_100
# `head` command writes to standard output a specified number of lines or bytes of each specified files
# to display the first 10 lines of a file
## head /usr/local/doc/vi.txt
# to display the first 5 lines of several files
## head -5 *.xdh
# or
## head -n 5 *.xdh

/content
--2024-04-03 06:02:40--  https://dl.fbaipublicfiles.com/XML/codes_xnli_100
Resolving dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)... 3.163.189.108, 3.163.189.51, 3.163.189.96, ...
Connecting to dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)|3.163.189.108|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2024-04-03 06:02:40 ERROR 403: Forbidden.

total 4
drwxr-xr-x 1 root root 4096 Apr  1 13:24 sample_data
head: cannot open 'codes_xnli_100' for reading: No such file or directory


### using `git clone` to clone git repository

In [ ]:
!git clone https://github.com/sally20921/running-tests-in-computer-vision

# Importing a library that is not in Colaboratory

To import a library that is not in Colaboratory by default, you can use `!pip install` or `!apt-get install`.


In [ ]:
# https://pypi.python.org/pypi/pydot
!apt-get -qq install -y graphviz && pip install pydot
import pydot

# Using Google Colab with GitHub

[https://colab.research.google.com/github/googlecolab/colabtools/blob/master/notebooks/colab-github-demo.ipynb](https://colab.research.google.com/github/googlecolab/colabtools/blob/master/notebooks/colab-github-demo.ipynb)

Google Colaboratory is designed to integrate cleanly with GitHub, allowing both loading notebooks from github and saving notebooks to github.


## Loading Public Notebooks Directly from GitHub

Colab can load public github notebooks directly, with no required authorization step.
